src/main/java/io/nats/client/impl/NatsJetStream.java

```diff
@@ -320,74 +342,141 @@ else if (so.isBind()) {
+            : PUSH_STATUS_MANAGER_FACTORY.createPushStatusMessageManager(conn, so, fnlServerCC, qgroup != null, dispatcher == null);
 
         NatsJetStreamSubscription sub;
-        if (dispatcher == null) {
-            sub = (NatsJetStreamSubscription) conn.createSubscription(inboxDeliver, qgroup, null, factory);
+        if (isPullMode) {
+            final NatsSubscriptionFactory factory = (sid, lSubject, lQgroup, lConn, lDispatcher)
+                -> new NatsJetStreamPullSubscription(sid, lSubject, lConn, this, fnlStream, fnlConsumerName, statusManager);
+            sub = (NatsJetStreamSubscription) conn.createSubscription(fnlInboxDeliver, qgroup, null, factory);
         }
         else {
-            MessageHandler handler;
-            if (autoAck && serverCC.getAckPolicy() != AckPolicy.None) {
-                handler = msg -> {
-                    if (asm.manage(msg)) { return; }  // manager handled the message
-                    userHandler.onMessage(msg);
-                    if (msg.lastAck() == null || msg.lastAck() == AckType.AckProgress) {
-                        msg.ack();
-                    }
-                };
+            final MessageManager sidCheckManager = so.isOrdered() ? new SidCheckManager() : null;
+
+            final MessageManager orderedManager = so.isOrdered()
+                ? new OrderedManager(this, dispatcher, fnlStream, fnlServerCC)
+                : null;
+
+            final NatsSubscriptionFactory factory = (sid, lSubject, lQgroup, lConn, lDispatcher)
+                -> new NatsJetStreamSubscription(sid, lSubject, lQgroup, lConn, lDispatcher,
+                this, fnlStream, fnlConsumerName, sidCheckManager, statusManager, orderedManager);
+
+            if (dispatcher == null) {
+                sub = (NatsJetStreamSubscription) conn.createSubscription(fnlInboxDeliver, qgroup, null, factory);
             }
             else {
-                handler = msg -> {
-                    if (asm.manage(msg)) { return; }  // manager handled the message
-                    userHandler.onMessage(msg);
-                };
+                AsyncMessageHandler handler = new AsyncMessageHandler(userHandler, isAutoAck, fnlServerCC, sidCheckManager, statusManager, orderedManager);
+                sub = (NatsJetStreamSubscription) dispatcher.subscribeImplJetStream(fnlInboxDeliver, qgroup, handler, factory);
             }
-            sub = (NatsJetStreamSubscription) dispatcher.subscribeImplJetStream(inboxDeliver, qgroup, handler, factory);
         }
 
-        asm.setSub(sub);
         return sub;
     }
 
+    static class ConsumerConfigurationComparer extends ConsumerConfiguration {
+        public ConsumerConfigurationComparer(ConsumerConfiguration cc) {
+            super(cc);
+        }
+
+        public boolean wouldBeChangeTo(ConsumerConfiguration serverCc) {
+            ConsumerConfigurationComparer serverCcc = new ConsumerConfigurationComparer(serverCc);
```

Comment:
implement equals and tohash instead?

Reply:
It's not the same as equals, because equals would understand "not" supplied to be a default value. I don't want this to be public at all, but it's the only way I can test it.

Reply:
This is not the same as equals, which would understand values not set to be equivalent to a default value. This wants to check if the user specifically set values intended as changes to an existing consumer